File: dotnet/src/Day4/Puzzle.cs
using System.Linq;

namespace AdventOfCode.Day4
{
    public static class Puzzle
    {
        public static bool SolvePart1(string passphrase)
        {
            var words = passphrase.Split(' ');
            var distinctWords = words.Distinct().ToArray();

            return words.Length == distinctWords.Length;
        }

        public static bool SolvePart2(string passphrase)
        {
            var words = passphrase
                .Split(' ')
                .Select(word => word.ToCharArray())
                .Select(charArray => charArray.OrderBy(character => character))
                .Select(charArray => new string(charArray.ToArray()))
                .ToArray();

            var distinctWords = words.Distinct().ToArray();

            return words.Length == distinctWords.Length;
        }
    }
}

File: go/src/day4/puzzle.go
package day4

import (
	"sort"
	"strings"
)

func SolvePart1(passphrase string) bool {
	words := map[string]bool{}

	for _, word := range strings.Split(passphrase, " ") {
		if words[word] == true {
			return false
		}

		words[word] = true
	}

	return true
}

func SolvePart2(passphrase string) bool {
	words := map[string]bool{}

	for _, word := range strings.Split(passphrase, " ") {
		runes := []rune(word)

		sort.Slice(runes, func(i, j int) bool {
			return runes[i] < runes[j]
		})

		word := string(runes)

		if words[word] == true {
			return false
		}

		words[word] = true
	}

	return true
}
